pkg/apis/autoscaling/validation/validation.go

```diff
@@ -44,8 +47,20 @@ var ValidateHorizontalPodAutoscalerName = apivalidation.ValidateReplicationContr
 
 func validateHorizontalPodAutoscalerSpec(autoscaler autoscaling.HorizontalPodAutoscalerSpec, fldPath *field.Path) field.ErrorList {
 	allErrs := field.ErrorList{}
-	if autoscaler.MinReplicas != nil && *autoscaler.MinReplicas < 1 {
-		allErrs = append(allErrs, field.Invalid(fldPath.Child("minReplicas"), *autoscaler.MinReplicas, "must be greater than 0"))
+
+	// MinReplicasLowerBound represents a minimum value for minReplicas
+	// 0 when HPA scale-to-zero feature is enabled
+	var MinReplicasLowerBound int32
+
+	if !utilfeature.DefaultFeatureGate.Enabled(features.HPAScaleToZero) {
+		MinReplicasLowerBound = 1
```

Comment:
validateHorizontalPodAutoscalerSpec is called from ValidateHorizontalPodAutoscalerUpdate.

Once we enable this feature by default (e.g. 1.17), an API server will allow minReplicas of 0. In a skewed HA apiserver scenario, the prior version of the apiserver with the feature disabled (e.g. 1.16) would reject updates to the object the 1.17 allowed to be persisted. If a 1.16 server encounters a persisted object that already has minReplicas set to 0, it must allow updates of that object with minReplicas=0.

One way to do this:
* pass in `minReplicasLowerBound` as a parameter to validateHorizontalPodAutoscalerSpec.
* when called from `ValidateHorizontalPodAutoscaler` (on creation), compute `minReplicasLowerBound` the way you do here (based only on the feature gate)
* when called from `ValidateHorizontalPodAutoscalerUpdate` (on update), allow a lower bound of 0 if the feature gate is enabled OR if the existing HPA has minReplicas=0.

Reply:
This also needs unit tests exercising all of these scenarios:
* feature enabled, ValidateHorizontalPodAutoscaler() allows minReplicas=0
* feature disabled, ValidateHorizontalPodAutoscaler() forbids minReplicas=0
* feature enabled, old HPA has minReplicas=1, ValidateHorizontalPodAutoscalerUpdate() allows minReplicas=0
* feature enabled, old HPA has minReplicas=0, ValidateHorizontalPodAutoscalerUpdate() allows minReplicas=0
* feature disabled, old HPA has minReplicas=1, ValidateHorizontalPodAutoscalerUpdate() forbids minReplicas=0
* feature disabled, old HPA has minReplicas=0, ValidateHorizontalPodAutoscalerUpdate() allows minReplicas=0

Reply:
@liggitt I've updated validation logic and moved `minReplicas` tests into separate [test functions](https://github.com/kubernetes/kubernetes/pull/74526/files#diff-f9431b410b931831536d0896b59f1c44R1083) to isolate feature gate toggling.